# Inspecting a DataFrame Object

## About the Data
In this notebook, we will be working with FIFA players data for 2022 obtained from [Kaggle](https://www.kaggle.com/datasets/stefanoleone992/fifa-22-complete-player-dataset)

## Setup
We will be working with the `players_22.csv` file, so we need to handle our imports and read it in.

In [19]:
import pandas as pd

In [3]:
players = pd.read_csv('players_22.csv')

/usr/local/lib/python3.8/dist-packages/IPython/core/interactiveshell.py:3326: DtypeWarning: Columns (26,109) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


## Examining dataframes
### Is it empty?

In [4]:
players.empty

False

### What are the dimensions?

In [5]:
players.shape

(19239, 111)

### What columns do we have?
We know there are 111 columns, but what are they? Let's use the `columns` attribute to see:

In [6]:
players.columns

Index(['sofifa_id', 'player_url', 'short_name', 'long_name',
       'player_positions', 'overall', 'potential', 'value_eur', 'wage_eur',
       'age',
       ...
       'lcb', 'cb', 'rcb', 'rb', 'gk', 'player_face_url', 'club_logo_url',
       'club_flag_url', 'nation_logo_url', 'nation_flag_url'],
      dtype='object', length=111)

### What does the data look like?
View rows from the top with `head()`:

In [7]:
players.head()

sofifa_id                                         player_url  \
0     158023  https://sofifa.com/player/158023/lionel-messi/...   
1     188545  https://sofifa.com/player/188545/robert-lewand...   
2      20801  https://sofifa.com/player/20801/c-ronaldo-dos-...   
3     190871  https://sofifa.com/player/190871/neymar-da-sil...   
4     192985  https://sofifa.com/player/192985/kevin-de-bruy...   

          short_name                            long_name player_positions  \
0           L. Messi       Lionel Andrés Messi Cuccittini       RW, ST, CF   
1     R. Lewandowski                   Robert Lewandowski               ST   
2  Cristiano Ronaldo  Cristiano Ronaldo dos Santos Aveiro           ST, LW   
3          Neymar Jr        Neymar da Silva Santos Júnior          LW, CAM   
4       K. De Bruyne                      Kevin De Bruyne          CM, CAM   

   overall  potential    value_eur  wage_eur  age  ...   lcb    cb   rcb  \
0       93         93   78000000.0  320000.0   34  ...  50+3  50+3  50+3   
1       92         92  119500000.0  270000.0   32  ...  60+3  60+3  60+3   
2       91         91   45000000.0  270000.0   36  ...  53+3  53+3  53+3   
3       91         91  129000000.0  270000.0   29  ...  50+3  50+3  50+3   
4       91         91  125500000.0  350000.0   30  ...  69+3  69+3  69+3   

     rb    gk                                    player_face_url  \
0  61+3  19+3  https://cdn.sofifa.net/players/158/023/22_120.png   
1  61+3  19+3  https://cdn.sofifa.net/players/188/545/22_120.png   
2  60+3  20+3  https://cdn.sofifa.net/players/020/801/22_120.png   
3  62+3  20+3  https://cdn.sofifa.net/players/190/871/22_120.png   
4  75+3  21+3  https://cdn.sofifa.net/players/192/985/22_120.png   

                            club_logo_url  \
0  https://cdn.sofifa.net/teams/73/60.png   
1  https://cdn.sofifa.net/teams/21/60.png   
2  https://cdn.sofifa.net/teams/11/60.png   
3  https://cdn.sofifa.net/teams/73/60.png   
4  https://cdn.sofifa.net/teams/10/60.png   

                             club_flag_url  \
0      https://cdn.sofifa.net/flags/fr.png   
1      https://cdn.sofifa.net/flags/de.png   
2  https://cdn.sofifa.net/flags/gb-eng.png   
3      https://cdn.sofifa.net/flags/fr.png   
4  https://cdn.sofifa.net/flags/gb-eng.png   

                            nation_logo_url  \
0  https://cdn.sofifa.net/teams/1369/60.png   
1  https://cdn.sofifa.net/teams/1353/60.png   
2  https://cdn.sofifa.net/teams/1354/60.png   
3                                       NaN   
4  https://cdn.sofifa.net/teams/1325/60.png   

                       nation_flag_url  
0  https://cdn.sofifa.net/flags/ar.png  
1  https://cdn.sofifa.net/flags/pl.png  
2  https://cdn.sofifa.net/flags/pt.png  
3  https://cdn.sofifa.net/flags/br.png  
4  https://cdn.sofifa.net/flags/be.png  

[5 rows x 111 columns]

View rows from the bottom with `tail()`. Let's view 2 rows:

In [8]:
players.tail(2)

sofifa_id                                         player_url  \
19237     262820  https://sofifa.com/player/262820/luke-rudden/2...   
19238     264540  https://sofifa.com/player/264540/emanuel-lalch...   

               short_name                long_name player_positions  overall  \
19237           L. Rudden              Luke Rudden               ST       47   
19238  E. Lalchhanchhuaha  Emanuel Lalchhanchhuaha              CAM       47   

       potential  value_eur  wage_eur  age  ...   lcb    cb   rcb    rb    gk  \
19237         60   110000.0     500.0   19  ...  26+2  26+2  26+2  32+2  15+2   
19238         60   110000.0     500.0   19  ...  41+2  41+2  41+2  45+2  16+2   

                                         player_face_url  \
19237  https://cdn.sofifa.net/players/262/820/22_120.png   
19238  https://cdn.sofifa.net/players/264/540/22_120.png   

                                    club_logo_url  \
19237  https://cdn.sofifa.net/teams/111131/60.png   
19238  https://cdn.sofifa.net/teams/113040/60.png   

                             club_flag_url nation_logo_url  \
19237  https://cdn.sofifa.net/flags/ie.png             NaN   
19238  https://cdn.sofifa.net/flags/in.png             NaN   

                           nation_flag_url  
19237  https://cdn.sofifa.net/flags/ie.png  
19238  https://cdn.sofifa.net/flags/in.png  

[2 rows x 111 columns]

### What datatypes do we have?

In [9]:
players.dtypes

sofifa_id            int64
player_url          object
short_name          object
long_name           object
player_positions    object
                     ...  
player_face_url     object
club_logo_url       object
club_flag_url       object
nation_logo_url     object
nation_flag_url     object
Length: 111, dtype: object

### Getting extra info and finding nulls

In [10]:
players.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19239 entries, 0 to 19238
Columns: 111 entries, sofifa_id to nation_flag_url
dtypes: float64(16), int64(44), object(51)
memory usage: 16.3+ MB


## Describing and Summarizing
### Get summary statistics

Note that `describe` only works for columns with numeric datatypes by default

In [11]:
players.describe()

sofifa_id       overall     potential     value_eur       wage_eur  \
count   19239.000000  19239.000000  19239.000000  1.916500e+04   19178.000000   
mean   231468.086959     65.772182     71.079370  2.850452e+06    9017.989363   
std     27039.717497      6.880232      6.086213  7.613700e+06   19470.176724   
min        41.000000     47.000000     49.000000  9.000000e+03     500.000000   
25%    214413.500000     61.000000     67.000000  4.750000e+05    1000.000000   
50%    236543.000000     66.000000     71.000000  9.750000e+05    3000.000000   
75%    253532.500000     70.000000     75.000000  2.000000e+06    8000.000000   
max    264640.000000     93.000000     95.000000  1.940000e+08  350000.000000   

                age     height_cm     weight_kg   club_team_id  league_level  \
count  19239.000000  19239.000000  19239.000000   19178.000000  19178.000000   
mean      25.210822    181.299704     74.943032   50580.498123      1.354364   
std        4.748235      6.863179      7.069434   54401.868535      0.747865   
min       16.000000    155.000000     49.000000       1.000000      1.000000   
25%       21.000000    176.000000     70.000000     479.000000      1.000000   
50%       25.000000    181.000000     75.000000    1938.000000      1.000000   
75%       29.000000    186.000000     80.000000  111139.000000      1.000000   
max       54.000000    206.000000    110.000000  115820.000000      5.000000   

       ...  mentality_composure  defending_marking_awareness  \
count  ...         19239.000000                 19239.000000   
mean   ...            57.929830                    46.601746   
std    ...            12.159326                    20.200807   
min    ...            12.000000                     4.000000   
25%    ...            50.000000                    29.000000   
50%    ...            59.000000                    52.000000   
75%    ...            66.000000                    63.000000   
max    ...            96.000000                    93.000000   

       defending_standing_tackle  defending_sliding_tackle  \
count               19239.000000              19239.000000   
mean                   48.045584                 45.906700   
std                    21.232718                 20.755683   
min                     5.000000                  5.000000   
25%                    28.000000                 25.000000   
50%                    56.000000                 53.000000   
75%                    65.000000                 63.000000   
max                    93.000000                 92.000000   

       goalkeeping_diving  goalkeeping_handling  goalkeeping_kicking  \
count        19239.000000          19239.000000         19239.000000   
mean            16.406102             16.192474            16.055356   
std             17.574028             16.839528            16.564554   
min              2.000000              2.000000             2.000000   
25%              8.000000              8.000000             8.000000   
50%             11.000000             11.000000            11.000000   
75%             14.000000             14.000000            14.000000   
max             91.000000             92.000000            93.000000   

       goalkeeping_positioning  goalkeeping_reflexes  goalkeeping_speed  
count             19239.000000          19239.000000        2132.000000  
mean                 16.229274             16.491814          36.439962  
std                  17.059779             17.884833          10.751563  
min                   2.000000              2.000000          15.000000  
25%                   8.000000              8.000000          27.000000  
50%                  11.000000             11.000000          36.000000  
75%                  14.000000             14.000000          45.000000  
max                  92.000000             90.000000          65.000000  

[8 rows x 60 columns]

Specifying the 5<sup>th</sup> and 95<sup>th</sup> percentile:

In [12]:
players.describe(percentiles=[0.05, 0.95])

sofifa_id       overall     potential     value_eur       wage_eur  \
count   19239.000000  19239.000000  19239.000000  1.916500e+04   19178.000000   
mean   231468.086959     65.772182     71.079370  2.850452e+06    9017.989363   
std     27039.717497      6.880232      6.086213  7.613700e+06   19470.176724   
min        41.000000     47.000000     49.000000  9.000000e+03     500.000000   
5%     184133.900000     54.000000     62.000000  1.800000e+05     500.000000   
50%    236543.000000     66.000000     71.000000  9.750000e+05    3000.000000   
95%    263046.100000     77.000000     82.000000  1.150000e+07   37150.000000   
max    264640.000000     93.000000     95.000000  1.940000e+08  350000.000000   

                age     height_cm     weight_kg   club_team_id  league_level  \
count  19239.000000  19239.000000  19239.000000   19178.000000  19178.000000   
mean      25.210822    181.299704     74.943032   50580.498123      1.354364   
std        4.748235      6.863179      7.069434   54401.868535      0.747865   
min       16.000000    155.000000     49.000000       1.000000      1.000000   
5%        18.000000    170.000000     64.000000      44.000000      1.000000   
50%       25.000000    181.000000     75.000000    1938.000000      1.000000   
95%       34.000000    193.000000     87.000000  113301.000000      3.000000   
max       54.000000    206.000000    110.000000  115820.000000      5.000000   

       ...  mentality_composure  defending_marking_awareness  \
count  ...         19239.000000                 19239.000000   
mean   ...            57.929830                    46.601746   
std    ...            12.159326                    20.200807   
min    ...            12.000000                     4.000000   
5%     ...            36.000000                    11.000000   
50%    ...            59.000000                    52.000000   
95%    ...            76.000000                    73.000000   
max    ...            96.000000                    93.000000   

       defending_standing_tackle  defending_sliding_tackle  \
count               19239.000000              19239.000000   
mean                   48.045584                 45.906700   
std                    21.232718                 20.755683   
min                     5.000000                  5.000000   
5%                     12.000000                 12.000000   
50%                    56.000000                 53.000000   
95%                    75.000000                 73.000000   
max                    93.000000                 92.000000   

       goalkeeping_diving  goalkeeping_handling  goalkeeping_kicking  \
count        19239.000000          19239.000000         19239.000000   
mean            16.406102             16.192474            16.055356   
std             17.574028             16.839528            16.564554   
min              2.000000              2.000000             2.000000   
5%               6.000000              6.000000             6.000000   
50%             11.000000             11.000000            11.000000   
95%             66.000000             63.000000            62.000000   
max             91.000000             92.000000            93.000000   

       goalkeeping_positioning  goalkeeping_reflexes  goalkeeping_speed  
count             19239.000000          19239.000000        2132.000000  
mean                 16.229274             16.491814          36.439962  
std                  17.059779             17.884833          10.751563  
min                   2.000000              2.000000          15.000000  
5%                    6.000000              6.000000          20.000000  
50%                  11.000000             11.000000          36.000000  
95%                  64.000000             67.000000          55.000000  
max                  92.000000             90.000000          65.000000  

[8 rows x 60 columns]

Describe all columns:

In [13]:
players.describe(include='all')

sofifa_id                                         player_url  \
count    19239.000000                                              19239   
unique            NaN                                              19239   
top               NaN  https://sofifa.com/player/158023/lionel-messi/...   
freq              NaN                                                  1   
mean    231468.086959                                                NaN   
std      27039.717497                                                NaN   
min         41.000000                                                NaN   
25%     214413.500000                                                NaN   
50%     236543.000000                                                NaN   
75%     253532.500000                                                NaN   
max     264640.000000                                                NaN   

          short_name        long_name player_positions       overall  \
count          19239            19239            19239  19239.000000   
unique         18145            19219              674           NaN   
top     J. Rodríguez  Ladislav Krejčí               CB           NaN   
freq              13                2             2423           NaN   
mean             NaN              NaN              NaN     65.772182   
std              NaN              NaN              NaN      6.880232   
min              NaN              NaN              NaN     47.000000   
25%              NaN              NaN              NaN     61.000000   
50%              NaN              NaN              NaN     66.000000   
75%              NaN              NaN              NaN     70.000000   
max              NaN              NaN              NaN     93.000000   

           potential     value_eur       wage_eur           age  ...    lcb  \
count   19239.000000  1.916500e+04   19178.000000  19239.000000  ...  19239   
unique           NaN           NaN            NaN           NaN  ...    259   
top              NaN           NaN            NaN           NaN  ...   64+2   
freq             NaN           NaN            NaN           NaN  ...    540   
mean       71.079370  2.850452e+06    9017.989363     25.210822  ...    NaN   
std         6.086213  7.613700e+06   19470.176724      4.748235  ...    NaN   
min        49.000000  9.000000e+03     500.000000     16.000000  ...    NaN   
25%        67.000000  4.750000e+05    1000.000000     21.000000  ...    NaN   
50%        71.000000  9.750000e+05    3000.000000     25.000000  ...    NaN   
75%        75.000000  2.000000e+06    8000.000000     29.000000  ...    NaN   
max        95.000000  1.940000e+08  350000.000000     54.000000  ...    NaN   

           cb    rcb     rb     gk  \
count   19239  19239  19239  19239   
unique    259    259    193    115   
top      64+2   64+2   62+2   16+2   
freq      540    540    767   3851   
mean      NaN    NaN    NaN    NaN   
std       NaN    NaN    NaN    NaN   
min       NaN    NaN    NaN    NaN   
25%       NaN    NaN    NaN    NaN   
50%       NaN    NaN    NaN    NaN   
75%       NaN    NaN    NaN    NaN   
max       NaN    NaN    NaN    NaN   

                                          player_face_url  \
count                                               19239   
unique                                              19239   
top     https://cdn.sofifa.net/players/158/023/22_120.png   
freq                                                    1   
mean                                                  NaN   
std                                                   NaN   
min                                                   NaN   
25%                                                   NaN   
50%                                                   NaN   
75%                                                   NaN   
max                                                   NaN   

                                 club_logo_url  \
count                                    1917

This works on specific columns also:

In [14]:
players.age.describe()

count    19239.000000
mean        25.210822
std          4.748235
min         16.000000
25%         21.000000
50%         25.000000
75%         29.000000
max         54.000000
Name: age, dtype: float64

We can also get the unique values in the `preferred_foot` column:

In [15]:
players.preferred_foot.unique()

array(['Left', 'Right'], dtype=object)

We can then use `value_counts()` to see how many of each unique value we have:

In [18]:
players.preferred_foot.value_counts()

Right    14674
Left      4565
Name: preferred_foot, dtype: int64